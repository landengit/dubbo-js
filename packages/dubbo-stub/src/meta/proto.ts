import { DescriptorProto } from 'ts-proto-descriptors/dist/google/protobuf/descriptor'
import ProtoField from './proto-field'
import ProtoEnum from './proto-enum'

export default class Proto {
  fields: ProtoField[] = []

  constructor(private protoDesc: DescriptorProto) {
    // debugger
  }

  get name() {
    return this.protoDesc.name
  }

  getEnums(): ProtoEnum[] {
    let allEnums = [
      ...this.protoDesc.enumType.map((item) => new ProtoEnum(item))
    ]

    return allEnums
  }

  getFields() {
    if (this.fields?.length > 0) {
      return this.fields
    } else {
      this.fields = this.protoDesc.field.map(
        (fieldDesc) => new ProtoField(fieldDesc)
      )
      return this.fields
    }
  }

  getNestedProtos(): Proto[] {
    // debugger
    return this.protoDesc.nestedType.map((item) => new Proto(item))
  }
}
