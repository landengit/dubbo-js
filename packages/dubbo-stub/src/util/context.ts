import { TypeMap } from './types'
import { Options } from './options'

/** Provides a parameter object for passing around the various context/config data. */
export interface Context {
  options: Options
  typeMap: TypeMap
  // todo dong 2022/11/21
  utils: any
}
