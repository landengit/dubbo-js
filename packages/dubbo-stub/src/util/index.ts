import ReadStream = NodeJS.ReadStream
export function readToBuffer(stream: ReadStream): Promise<Buffer> {
  return new Promise((resolve) => {
    const ret: Array<Buffer | string> = []
    let len = 0
    stream.on('readable', () => {
      let chunk
      while ((chunk = stream.read())) {
        ret.push(chunk)
        len += chunk.length
      }
    })
    stream.on('end', () => {
      resolve(Buffer.concat(ret as any, len))
    })
  })
}
