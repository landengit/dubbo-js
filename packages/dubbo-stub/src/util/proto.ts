import { CodeGeneratorRequest } from 'ts-proto-descriptors/dist/google/protobuf/compiler/plugin'
import { readJsonFile, writeJsonFile } from 'nx/src/utils/fileutils'

export function dumpCodeGenRequest(
  path: string,
  request: CodeGeneratorRequest
) {
  writeJsonFile(path, CodeGeneratorRequest.toJSON(request) as any, {
    spaces: 2
  })
}

/**
 * load buffer from file
 * @param jsonPath
 */
export function loadCodeGenRequest(jsonPath: string): CodeGeneratorRequest {
  let jsonRequest = readJsonFile(jsonPath)
  return CodeGeneratorRequest.fromJSON(jsonRequest)
}
