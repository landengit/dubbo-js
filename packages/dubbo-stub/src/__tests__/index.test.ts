/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { loadCodeGenRequest } from '../util/proto'
import { join } from 'path'
import File from '../meta/FIle'
import ClientStubService from '../scene/client-stub-service'
// import { expect } from 'vitest'

describe(`basic`, () => {
  it('client demo gen', () => {
    const request = loadCodeGenRequest(
      join(__dirname, '../../src/__tests__/protobuf/DemoService.json')
    )
    const filesToGenerate = request.protoFile.map((file) => new File(file))
    for (const file of filesToGenerate) {
      let serviceItem = new ClientStubService(file)
      expect(serviceItem.getFilePath()).toMatchInlineSnapshot('"demoservice"')
      expect(serviceItem.genCode()).toMatchInlineSnapshot(`
        "export interface HelloRequest {
            name:string
        }

        export interface HelloReply {
            message:string
        }

        // generate service interface
        export interface IDemoService {
        SayHello(helloRequest: HelloRequest): Promise<HelloReply>
        }"
      `)
    }
  })

  it('all demo genTest', () => {
    let allFile = [
      'DemoService.json', //ok
      'health.json', //ok
      'DashAPICreds.json', //ok
      'reflection.json' //ok anyone 未处理
    ]

    for (const fileName of allFile) {
      const request = loadCodeGenRequest(
        join(__dirname, '../../src/__tests__/protobuf/' + fileName)
      )
      const filesToGenerate = request.protoFile.map((file) => new File(file))
      for (const file of filesToGenerate) {
        let serviceItem = new ClientStubService(file)
        expect(serviceItem.genCode()).toMatchSnapshot(fileName)
      }
    }
  })
})
