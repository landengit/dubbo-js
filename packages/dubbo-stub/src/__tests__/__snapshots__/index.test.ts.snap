// Vitest Snapshot v1

exports[`basic > all demo genTest > DashAPICreds.json 1`] = `
"export interface DashFlash {
    msg:string

type:Type
}

export interface DashUserSettingsState {
    email:string

urls:URLs

flashes:DashFlash
}

export interface DashCred {
    description:string

metadata:string

token:string

id:string
}

export interface DashAPICredsCreateReq {
    description:string

metadata:string
}

export interface DashAPICredsUpdateReq {
    cred_sid:string

description:string

metadata:string

id:string
}

export interface DashAPICredsDeleteReq {
    cred_sid:string

id:string
}

export interface Empty {
    
}

export interface URLs {
    connect_google:string

connect_github:string
}

 export enum Type {
  UNKNOWN = 0,
  SERVING = 1,
  NOT_SERVING = 2,
  SERVICE_UNKNOWN = 3
}


// generate service interface
export interface IDashState {
UserSettings(empty: Empty): Promise<DashUserSettingsState>

ActiveUserSettingsStream(empty: Empty): Promise<DashUserSettingsState>

ChangeUserSettingsStream(dashUserSettingsState: DashUserSettingsState): Promise<DashUserSettingsState>
}

// generate service interface
export interface IDashAPICreds {
Create(dashAPICredsCreateReq: DashAPICredsCreateReq): Promise<DashCred>

Update(dashAPICredsUpdateReq: DashAPICredsUpdateReq): Promise<DashCred>

Delete(dashAPICredsDeleteReq: DashAPICredsDeleteReq): Promise<DashCred>
}"
`;

exports[`basic > all demo genTest > DemoService.json 1`] = `
"export interface HelloRequest {
    name:string
}

export interface HelloReply {
    message:string
}

// generate service interface
export interface IDemoService {
SayHello(helloRequest: HelloRequest): Promise<HelloReply>
}"
`;

exports[`basic > all demo genTest > health.json 1`] = `
"export interface HealthCheckRequest {
    service:string
}

export interface HealthCheckResponse {
    status:ServingStatus
}

 export enum ServingStatus {
  UNKNOWN = 0,
  SERVING = 1,
  NOT_SERVING = 2,
  SERVICE_UNKNOWN = 3
}


// generate service interface
export interface IHealth {
Check(healthCheckRequest: HealthCheckRequest): Promise<HealthCheckResponse>

Watch(healthCheckRequest: HealthCheckRequest): Promise<HealthCheckResponse>
}"
`;

exports[`basic > all demo genTest > reflection.json 1`] = `
"export interface ServerReflectionRequest {
    host:string

file_by_filename:string

file_containing_symbol:string

file_containing_extension:ExtensionRequest

all_extension_numbers_of_type:string

list_services:string
}

export interface ExtensionRequest {
    containing_type:string

extension_number:number
}

export interface ServerReflectionResponse {
    valid_host:string

original_request:ServerReflectionRequest

file_descriptor_response:FileDescriptorResponse

all_extension_numbers_response:ExtensionNumberResponse

list_services_response:ListServiceResponse

error_response:ErrorResponse
}

export interface FileDescriptorResponse {
    file_descriptor_proto:Uint8Array
}

export interface ExtensionNumberResponse {
    base_type_name:string

extension_number:number
}

export interface ListServiceResponse {
    service:ServiceResponse
}

export interface ServiceResponse {
    name:string
}

export interface ErrorResponse {
    error_code:number

error_message:string
}

// generate service interface
export interface IServerReflection {
ServerReflectionInfo(serverReflectionRequest: ServerReflectionRequest): Promise<ServerReflectionResponse>
}"
`;
